### PR TITLE
#341/new tests for linkernamespace

### PR DESCRIPTION
**Pull request by tomuben**

### CODE REVIEW

File: exaudfclient/base/BUILD [hunk 1]
```diff
@@ -125,6 +125,39 @@ cc_binary(
     data = [":libexaudflib_complete.so"]
 )
 
+## The purpose of the static binaries is to verify if the linker namespace test work correctly.
+## It is crucial that all linker symbols of exaudflib and all it's dependency are loaded into a new linker namespace,
+## so that the user can load it's own versions of those dependencies without any conflict.
+## With this binary we simulate an error in our build system, that is a direct depdency to protobuf/zmq,
```
Comment (tkilias) [inline exaudfclient/base/BUILD:131]:
```suggestion
## With this binary we simulate an error in our build system, that is a direct depedency to protobuf/zmq,
```

File: exaudfclient/base/BUILD [hunk 2]
```diff
@@ -125,6 +125,39 @@ cc_binary(
...
+## which then must be detected with the linker namespace tests:
+## test/linker-namespace-sanity/linker_namespace_sanity.py checks the wrong configuration
+## Besides this the test under test/python3/all/linker_namespace.py checks the normal build, which expects
+## not to find any occurence of the dependencies (protobuf/zmq) in the primary linker namespace.
+##
+## We need to explicitly declare the dependency of protobuf/zmq here, as the exaudflib is a static lib (//exaudflib:exaudflib)
+## an hence does not contain dependency information. We cannot declare the shared lib (:exaudflib_complete.so)
```
Comment (tkilias) [inline exaudfclient/base/BUILD:138]:
```suggestion
## and hence does not contain dependency information. We cannot declare the shared lib (:exaudflib_complete.so)
```

File: tests/test/generic/linker_namespace1.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python2.7
```
Comment (tkilias) [inline tests/test/generic/linker_namespace1.py:1]: Is this file used somewhere?

**Review by tkilias — CHANGES_REQUESTED** [review]
And, I saw the following test failed. I guess it is related to my question, if tests/test/generic/linker_namespace1.py is used somewhere.
```

INFO: RunDBTest_b8d1886e2b(db_user=sys, flavor_path=flavors/python-3.6-data-science-EXASOL-6.2.0, test_restrictions=[], release_goal=release, language=python3, test_file=generic/linker_namespace1.py): Running db tests
ERROR: RunDBTest_b8d1886e2b(db_user=sys, flavor_path=flavors/python-3.6-data-science-EXASOL-6.2.0, test_restrictions=[], release_goal=release, language=python3, test_file=generic/linker_namespace1.py): db tests failed
Test results:
command: bash -c "cd /tests/test/; python -tt "generic/linker_namespace1.py" --server '172.18.0.2:8888' --user 'sys' --password 'exasol' --script-languages 'PYTHON3=localzmq+protobuf:///bfsdefault/myudfs/python-3.6-data-science-EXASOL-6.2.0?lang=python#buckets/bfsdefault/myudfs/python-3.6-data-science-EXASOL-6.2.0/exaudf/exaudfclient_py3 PYTHON=builtin_python JAVA=builtin_java R=builtin_r' --loglevel=critical --driver=/downloads/ODBC/lib/linux/x86_64/libexaodbc-uo2214lv2.so  --jdbc-path /downloads/JDBC/exajdbc.jar --lang python3 " 
environment: {'TRAVIS': '', 'TEST_ENVIRONMENT_TYPE': 'docker_db', 'TEST_ENVIRONMENT_NAME': 'python-3.6-data-science-EXASOL-6.2.0_release', 'TEST_DOCKER_NETWORK_NAME': 'db_network_python-3.6-data-science-EXASOL-6.2.0_release', 'TEST_DOCKER_DB_CONTAINER_NAME': 'db_container_python-3.6-data-science-EXASOL-6.2.0_release'}
  File "generic/linker_namespace1.py", line 44
    def findSymbols(self, filter_strings : List[str]) -> List[str]:
                                         ^
SyntaxError: invalid syntax

```

File: tests/test/generic/linker_namespace1.py (hunk 3 above)
Comment (tomuben) [reply tests/test/generic/linker_namespace1.py]: I removed it, it was a debug file, and not supposed to get integrated.

File: exaudfclient/base/.bazelrc [hunk 4]
```diff
@@ -8,7 +8,14 @@ build:fast-binary-py3 --copt='-DCUSTOM_LIBEXAUDFLIB_PATH="/exaudf/libexaudflib_c
 build:fast-binary-both --config=fast-binary-py2 --config=fast-binary-py3
 build:slow-wrapper-py2 --define binary_type=slow_wrapper //:exaudfclient
 build:slow-wrapper-py3 --define binary_type=slow_wrapper //:exaudfclient_py3
-build:slow-wrapper-both --config=slow-wrapper-py2 --config=slow-wrapper-py3 
+build:slow-wrapper-both --config=slow-wrapper-py2 --config=slow-wrapper-py3
+build:static-binary-py3 //:exaudfclient_py3_static
+build:static-binary-py2 //:exaudfclient_py2_static
+#Static binary uses the same executable as the normal exaudflclient (with dynamic linking)
```
Comment (tkilias) [inline exaudfclient/base/.bazelrc:14]: Not sure if understand this comment

File: tests/test/linker-namespace-sanity/linker_namespace_sanity.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    @skipIf('PYTHON3' not in get_supported_languages(), "UDF does not support Python3")
+    def test_linker_namespace(self):
+
+        lang = getScriptLanguagesFromArgs()
+        r_py2 = re.compile(r"exaudf/exaudfclient\b") # Match "...exaudf/exaudfclient ..."/"...exaudf/exaudfclient" but not "...exaudf/exaudfclient_py3"
+        lang_static = r_py2.sub("exaudf/exaudfclient_py2_static", lang)
+        r_py3 = re.compile(r"exaudf/exaudfclient_py3\b")
+        lang_static = r_py3.sub("exaudf/exaudfclient_py3_static", lang_static)
+        alter_session_query_str = "ALTER SESSION SET SCRIPT_LANGUAGES='%s'" % lang_static
+        print(alter_session_query_str)
```
Comment (tkilias) [inline tests/test/linker-namespace-sanity/linker_namespace_sanity.py:31]: maybe, move into own function and give it a proper name

File: tests/test/linker-namespace-sanity/linker_namespace_base_test.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    def _execute_linker_namespace_udf(self, libs_to_check):
+        self.query(udf.fixindent('''
+                    CREATE OR REPLACE python3 SCALAR SCRIPT linker_namespace_test(search_string VARCHAR(100000)) RETURNS VARCHAR(100000) AS
+                    from ctypes import *
+                    from typing import List
+
+                    #Wrapper class for accessing libc function dl_iterate_phdr 
+                    #which is used to get all shared libraries in link-map (linker namespace)
+                    #See https://linux.die.net/man/3/dl_iterate_phdr for details of this function
+                    class SymbolScanner:
```
Comment (tkilias) [inline tests/test/linker-namespace-sanity/linker_namespace_base_test.py:26]: I am thinking about, if we maybe should move this to an extra python file which we read and add to the create script via string operations.

**Review by tkilias — CHANGES_REQUESTED** [review]
A few small things

File: exaudfclient/base/.bazelrc (hunk 4 above)
Comment (tomuben) [reply exaudfclient/base/.bazelrc:14]: I believe this comment is an artifact from a previous try. I removed it.

File: tests/test/linker-namespace-sanity/linker_namespace_base_test.py (hunk 6 above)
Comment (tomuben) [reply tests/test/linker-namespace-sanity/linker_namespace_base_test.py:26]: Hm, let's discuss offline.

File: tests/test/linker-namespace-sanity/linker_namespace_sanity.py (hunk 5 above)
Comment (tomuben) [reply tests/test/linker-namespace-sanity/linker_namespace_sanity.py:34]: Done

File: tests/test/linker-namespace-sanity/linker_namespace_base_test.py (hunk 6 above)
Comment (tomuben) [reply tests/test/linker-namespace-sanity/linker_namespace_base_test.py]: Done

File: exaudfclient/base/.bazelrc (hunk 4 above)
Comment (tomuben) [reply exaudfclient/base/.bazelrc]: Done

**Review by tkilias — APPROVED** [review]

**Pull request merged**